stage1_fly/run/main.go

```diff
@@ -199,19 +200,40 @@ func evaluateMounts(rfs string, app string, p *stage1commontypes.Pod) ([]flyMoun
 			)
 		}
 
-		var flags uintptr = syscall.MS_BIND
+		var (
+			flags     uintptr = syscall.MS_BIND
+			recursive         = tuple.V.Recursive != nil && *tuple.V.Recursive
+			ro                = tuple.V.ReadOnly != nil && *tuple.V.ReadOnly
+		)
+
 		// If Recursive is not set, default to non-recursive.
-		if tuple.V.Recursive != nil && *tuple.V.Recursive {
+		if recursive {
 			flags |= syscall.MS_REC
 		}
+
 		argFlyMounts = append(argFlyMounts,
 			flyMount{tuple.V.Source, rfs, tuple.M.Path, "none", flags},
 		)
 
-		if tuple.V.ReadOnly != nil && *tuple.V.ReadOnly {
+		if ro {
 			argFlyMounts = append(argFlyMounts,
 				flyMount{"", rfs, tuple.M.Path, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
 			)
+
+			if recursive {
+				// Every sub-mount needs to be remounted read-only separately
+				mnts, err := stage0.GetMountsForPrefix(tuple.V.Source + "/")
+				if err != nil {
+					return nil, errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", tuple.V.Source), err)
+				}
+
+				for _, mnt := range mnts {
+					innerRelPath := tuple.M.Path + strings.Replace(mnt.MountPoint, tuple.V.Source, "", -1)
+					argFlyMounts = append(argFlyMounts,
+						flyMount{"", rfs, innerRelPath, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
```

Comment:
> I don't think so: the "sb" in do_remount_sb means superblock, and the superblock is the data structure which is common to all mounts of the same underlying storage.

This is beyond my knowledge about mounts and I wonder how this relates to shared mounts. Do shared mounts automatically shared the same superblock? What happens if a shared mount gets remounted private or unshared?

Besides, seems like in all our cases we need MS_BIND to do achieve what we want.